syntax = "proto3";

package ovms_ipc;

import "ie_serving/messaging/apis/data_attributes.proto";

message EndpointResponse{
    oneof type{
        PredictResponse predict_response = 1;
        ModelMetadataResponse model_metadata_response = 2;
        ModelStatusResponse model_status_response = 3;
        ErrorResponse error_response = 4;
    }
}

message PredictResponse{
    message Data{
        oneof type{
            NumpyAttributes numpy_attributes = 1;
            StringAttributes string_attributes = 2;
            BinaryAttributes binary_attributes = 3;
        }
        string output_name = 4;
        string shm_name = 5;
    }
    repeated Data outputs = 1;
    int32 responding_version = 2;
}

message ModelMetadataResponse{
    message Layer{
        string layer_name = 1;
        string precision = 2;
        repeated int32 shape = 3;
    }
    message Signature{
        string name = 1;
        Layer layer = 2;
    }
    repeated Signature inputs_signature = 1;
    repeated Signature outputs_signature = 2;
    string method_name = 3;
}

message ModelStatusResponse{
    enum State{
        START = 0;
        LOADING = 1;
        AVAILABLE = 2;
        UNLOADING = 3;
        END = 4;
    }
    message Status{
        enum ErrorCode{
            OK = 0;
            UNKNOWN = 1;
        }
        ErrorCode error_code = 1;
        string error_message = 2;
    }
    message ModelVersionStatus{
        string model_name = 1;
        int32 model_version = 2;
        State state = 3;
        Status status = 4;
    }
}

message ErrorResponse{
    int32 error_code = 1;
    string error_message = 2;
}
